README.rst
=============================
Django Oauth Clients 3
=============================

.. image:: https://badge.fury.io/py/dj-oauth-clients3.svg
    :target: https://badge.fury.io/py/dj-oauth-clients3

.. image:: https://travis-ci.org/kverdecia/dj-oauth-clients3.svg?branch=master
    :target: https://travis-ci.org/kverdecia/dj-oauth-clients3

.. image:: https://codecov.io/gh/kverdecia/dj-oauth-clients3/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/kverdecia/dj-oauth-clients3

Django app for handling oauth2 clients (without support for python2 and django1.x)

Documentation
-------------

The full documentation is at https://dj-oauth-clients3.readthedocs.io.

Quickstart
----------

Install Django Oauth Clients 3::

    pip install dj-oauth-clients3

Add it to your `INSTALLED_APPS`:

.. code-block:: python

    INSTALLED_APPS = (
        ...
        'oauth_clients3.apps.OauthClients3Config',
        ...
    )

Add Django Oauth Clients 3's URL patterns:

.. code-block:: python

    from oauth_clients3 import urls as oauth_clients3_urls


    urlpatterns = [
        ...
        url(r'^', include(oauth_clients3_urls)),
        ...
    ]

Features
--------

* TODO

Running Tests
-------------

Does the code actually work?

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install tox
    (myenv) $ tox

Credits
-------

Tools used in rendering this package:

*  Cookiecutter_
*  `cookiecutter-djangopackage`_

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`cookiecutter-djangopackage`: https://github.com/pydanny/cookiecutter-djangopackage


HISTORY.rst
.. :changelog:

History
-------

0.1.6 (2019-12-02)
++++++++++++++++++

* First release on PyPI.


tests/urls.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

from django.conf.urls import url, include


urlpatterns = [
    url(r'^', include('oauth_clients3.urls', namespace='oauth_clients3')),
]


tests/settings.py
# -*- coding: utf-8
from __future__ import unicode_literals, absolute_import

import django

DEBUG = True
USE_TZ = True

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "!i)%7yzn*zqo^i-1i9i9@9es(c_84#=+g5phmu6y+f*$@#x#b*"

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.sqlite3",
        "NAME": ":memory:",
    }
}

ROOT_URLCONF = "tests.urls"

INSTALLED_APPS = [
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sites",
    "oauth_clients3",
]

SITE_ID = 1

if django.VERSION >= (1, 10):
    MIDDLEWARE = ()
else:
    MIDDLEWARE_CLASSES = ()


tests/test_models.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_dj-oauth-clients3
------------

Tests for `dj-oauth-clients3` models module.
"""

from django.test import TestCase

from oauth_clients3 import models


class TestOauth_clients3(TestCase):

    def setUp(self):
        pass

    def test_something(self):
        pass

    def tearDown(self):
        pass


oauth_clients3/admin.py
"Admin interface for defining and authorizing oauth2 clients."
from django.contrib import admin
from django.utils.translation import ugettext_lazy as _, ugettext
from django.utils.formats import localize
from django.conf.urls import url
from django.shortcuts import get_object_or_404, redirect
from django.contrib import messages
from . import models


class TokenInline(admin.TabularInline):
    "Inline admin interface for oauth2 client's access tokens"
    fields = ('user_id', 'username', 'token_type', 'expires_in', 'access_token', 'refresh_token')
    readonly_fields = fields
    extra = 0
    model = models.AccessToken

class ParamInline(admin.TabularInline):
    "Inline admin interface for oauth2 client's extra parameters."
    fields = ('name', 'value')
    extra = 2
    model = models.ClientParam

class ClientAdmin(admin.ModelAdmin):
    "Admin interface for managing oauth2 clients."
    list_display = ('uid', 'name', 'client_id', 'authorization_endpoint', 'token_endpoint',
        'scope', 'creator', 'created', 'modified')
    fields = ('uid', 'name', 'client_id', 'client_secret', 'authorization_endpoint',
        'token_endpoint', 'creator', 'scope', 'created', 'modified', 'complete_authorization_url')
    readonly_fields = ('uid', 'creator', 'created', 'modified', 'complete_authorization_url')
    inlines = [ParamInline, TokenInline]

    def get_urls(self):
        urls = super(ClientAdmin, self).get_urls()
        info = self.model._meta.app_label, self.model._meta.model_name
        my_urls = [
            url(r'^(.+)/complete-authorization/$', self.complete_authorization, name='%s_%s_complete_authorization' % info),
        ]
        return my_urls + urls

    def complete_authorization(self, request, uid):
        "View to complete oauth2 authorization"
        client = get_object_or_404(models.Client, uid=uid)
        redirect_url = client.complete_authorization_url(request)
        final_redirection = client.complete_authorization(request, redirect_url)
        msg = ugettext("Oauth client authorized.")
        self.message_user(request, msg, messages.INFO)
        if final_redirection:
            return redirect(final_redirection)
        url_name = 'admin:{}_{}_change'.format(self.model._meta.app_label, self.model._meta.model_name)
        return redirect(url_name, client.pk)

    def save_model(self, request, obj, form, change):
        if obj.pk is None:
            obj.creator = request.user
        super(ClientAdmin, self).save_model(request, obj, form, change)

    def response_change(self, request, obj):
        if '_authorize-client' in request.POST:
            redirect_url = obj.complete_authorization_url(request)
            authorization_url = obj.start_authorization_url(request, redirect_url)
            return redirect(authorization_url)
        return super(ClientAdmin, self).response_change(request, obj)


class TokenAdmin(admin.ModelAdmin):
    "Admin interface for managing oauth2 client access tokens."
    list_display = ('client', 'user_id', 'username', 'token_type', 'get_expiration',
        'get_expired', 'modified')
    fields = ('client', 'user_id', 'username', 'token_type', 'scope', 'expires_in',
        'get_expiration', 'get_expired', 'access_token', 'refresh_token', 'created',
        'modified')
    readonly_fields = fields
    list_filter = ('client', 'created', 'modified')
    actions = ['refresh_tokens']

    def has_add_permission(self, request):
        return False

    def refresh_tokens(self, request, queryset):
        "Action for refreshing tokens."
        for access_token in queryset:
            access_token.do_refresh_token()
        msg = ugettext("Selected tokens refreshed.")
        self.message_user(request, msg, messages.INFO)
    refresh_tokens.short_description = _("Refresh selected tokens")

    def get_expiration(self, obj):
        "Column for showing the expiration date."
        return localize(obj.expiration())
    get_expiration.short_description = _("Expiration")

    def get_expired(self, obj):
        "Column for showing if the toke is expired."
        return obj.is_expired()
    get_expired.short_description = _("Expired")
    get_expired.boolean = True


admin.site.register(models.Client, ClientAdmin)
admin.site.register(models.AccessToken, TokenAdmin)


oauth_clients3/migrations/0001_initial.py
# Generated by Django 3.0 on 2019-12-02 21:39

from django.conf import settings
from django.db import migrations, models
import django.db.models.deletion
import uuid


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
    ]

    operations = [
        migrations.CreateModel(
            name='Client',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True, verbose_name='Created')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='Modified')),
                ('uid', models.UUIDField(default=uuid.uuid4, unique=True, verbose_name='UID')),
                ('name', models.CharField(max_length=100, unique=True, verbose_name='Name')),
                ('client_id', models.CharField(max_length=255, unique=True, verbose_name='Client id')),
                ('client_secret', models.CharField(max_length=255, verbose_name='Client secret')),
                ('authorization_endpoint', models.URLField(max_length=255, verbose_name='Authorization endpoint')),
                ('token_endpoint', models.URLField(blank=True, default='', max_length=255, verbose_name='Token endpoint')),
                ('scope', models.CharField(blank=True, default='read', max_length=255, verbose_name='Scope')),
                ('creator', models.ForeignKey(on_delete=django.db.models.deletion.PROTECT, to=settings.AUTH_USER_MODEL, verbose_name='User')),
            ],
            options={
                'verbose_name': 'Oauth2 client',
                'verbose_name_plural': 'Oauth2 clients',
            },
        ),
        migrations.CreateModel(
            name='ClientParam',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=40, verbose_name='Name')),
                ('value', models.TextField(blank=True, default='', verbose_name='Value')),
                ('client', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, related_name='params', to='oauth_clients3.Client', verbose_name='Client')),
            ],
            options={
                'verbose_name': 'Client param',
                'verbose_name_plural': 'Client params',
            },
        ),
        migrations.CreateModel(
            name='AccessToken',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(auto_now_add=True, verbose_name='Created')),
                ('modified', models.DateTimeField(auto_now=True, verbose_name='Modified')),
                ('user_id', models.CharField(blank=True, default='', max_length=255, verbose_name='User id')),
                ('username', models.CharField(blank=True, default='', max_length=255, verbose_name='Username')),
                ('token_type', models.CharField(max_length=255, verbose_name='Token type')),
                ('expires_in', models.IntegerField(verbose_name='Expires in')),
                ('access_token', models.CharField(max_length=255, verbose_name='Access token')),
                ('refresh_token', models.CharField(max_length=255, verbose_name='Refresh token')),
                ('scope', models.TextField(blank=True, default='', verbose_name='Scope')),
                ('client', models.ForeignKey(on_delete=django.db.models.deletion.CASCADE, to='oauth_clients3.Client', verbose_name='Client')),
            ],
            options={
                'verbose_name': 'Access token',
                'verbose_name_plural': 'Access tokens',
            },
        ),
    ]


oauth_clients3/models.py
"Models for managing oath2 client authorizations."
# -*- coding: utf-8 -*-
import uuid
import datetime
from urllib.parse import urlencode
import requests
from django.conf import settings
from django.db import models
from django.utils import timezone
from django.utils.translation import ugettext_lazy as _
from django.urls import reverse


class TimeStampedModel(models.Model):
    """
    An abstract base class model that provides self-updating
    ``created`` and ``modified`` fields.
    """
    created = models.DateTimeField(_('Created'), auto_now_add=True, editable=False)
    modified = models.DateTimeField(_('Modified'), auto_now=True, editable=False)

    class Meta:
        abstract = True

class Client(TimeStampedModel):
    "Configuration and authentication of an oauth2 clients."
    uid = models.UUIDField(_("UID"), default=uuid.uuid4, unique=True)
    name = models.CharField(_("Name"), max_length=100, unique=True)
    client_id = models.CharField(_("Client id"), max_length=255, unique=True)
    client_secret = models.CharField(_("Client secret"), max_length=255)
    authorization_endpoint = models.URLField(_("Authorization endpoint"), max_length=255)
    token_endpoint = models.URLField(_("Token endpoint"), max_length=255, blank=True, default='')
    creator = models.ForeignKey(settings.AUTH_USER_MODEL, verbose_name=_("User"), on_delete=models.PROTECT)
    scope = models.CharField(_("Scope"), max_length=255, blank=True, default='read')

    class Meta:
        verbose_name = _("Oauth2 client")
        verbose_name_plural = _("Oauth2 clients")

    def __str__(self):
        return self.name

    def save(self, *args, **kwargs):
        "Saves the client."
        if not self.token_endpoint:
            self.token_endpoint = self.authorization_endpoint
        return super(Client, self).save(*args, **kwargs)

    @property
    def session_state_name(self):
        "Returns the id of the oauth2 authentication process session."
        return str(self.uid)

    @property
    def session_redirect_url_name(self):
        "Returns the name of the redirection view."
        return '{}:redirect-url'.format(str(self.uid))

    def complete_authorization_url(self, request=None):
        "Returns the redirection url."
        info = self._meta.app_label, self._meta.model_name
        view_name = 'admin:%s_%s_complete_authorization' % info
        authorization_url = reverse(view_name, args=(self.uid,))
        return request.build_absolute_uri(authorization_url) if request else authorization_url

    def start_authorization_url(self, request, redirect_url, final_redirection=None):
        """Returns the oauth2 provider url that will be redirected to when starting
        the authorization process.

        Parameters
        ----------
        request
            Django request object
        redirect_url: string
            Redirection url that will be sent to the authorization server.
        final_redirection: string
            If provided, after the redirection from the oauth2 provider, the
            user will be redirected to the url in this parameter.

        Returns
        -------
        url: string
            Returns the url to use to start the authorization process.
        """
        redirect_url = request.build_absolute_uri(redirect_url)
        state = str(uuid.uuid4())
        params = {
            'response_type': 'code',
            'client_id': self.client_id,
            'redirect_uri': redirect_url,
            'scope': self.scope,
            'state': state,
        }
        for row in self.params.all():
            params[row.name] = row.value
        encoded_params = urlencode(params)
        result = '{}?{}'.format(self.authorization_endpoint, encoded_params)
        request.session[self.session_state_name] = state
        if final_redirection:
            request.session[self.session_redirect_url_name] = final_redirection
        return result

    def complete_authorization(self, request, redirect_url):
        """Execute this method in the redirection view to fetch the access
        token."""
        state = request.GET['state']
        if request.session[self.session_state_name] != state:
            raise ValueError("Wrong oauth state.")
        del request.session[self.session_state_name]
        code = request.GET['code']
        token_url = '{}?grant_type=authorization_code&client_id={}&client_secret={}&redirect_uri={}&code={}'
        token_url = token_url.format(self.token_endpoint, self.client_id,
            self.client_secret, redirect_url, code)
        response = requests.post(token_url)
        data = response.json()
        access_token = AccessToken()
        access_token.client = self
        access_token.token_type = data['token_type']
        access_token.expires_in = data['expires_in']
        access_token.access_token = data['access_token']
        access_token.refresh_token = data.get('refresh_token', '')
        access_token.scope = data.get('scope', self.scope)
        access_token.save()
        return request.session.pop(self.session_redirect_url_name, None)

class ClientParam(models.Model):
    "Define aditional params to pass the start authorization url."
    client = models.ForeignKey(Client, verbose_name=_("Client"),
        related_name='params', on_delete=models.CASCADE)
    name = models.CharField(_("Name"), max_length=40)
    value = models.TextField(_("Value"), blank=True, default='')

    class Meta:
        verbose_name = _("Client param")
        verbose_name_plural = _("Client params")

    def __str__(self):
        return self.name


class AccessToken(TimeStampedModel):
    """Oauth2 client access token. This object shold be usually generated
    from the oauth2 client authorization process."""
    client = models.ForeignKey(Client, verbose_name=_("Client"), on_delete=models.CASCADE)
    user_id = models.CharField(_("User id"), max_length=255, blank=True, default='')
    username = models.CharField(_("Username"), max_length=255, blank=True, default='')
    token_type = models.CharField(_("Token type"), max_length=255)
    expires_in = models.IntegerField(_("Expires in"))
    access_token = models.CharField(_("Access token"), max_length=255)
    refresh_token = models.CharField(_("Refresh token"), max_length=255)
    scope = models.TextField(_("Scope"), blank=True, default='')

    class Meta:
        verbose_name = _("Access token")
        verbose_name_plural = _("Access tokens")    

    def do_refresh_token(self):
        "Refresh the token"
        url = '{}?grant_type=refresh_token&refresh_token={}&client_id={}&client_secret={}&scope={}'
        url = url.format(self.client.token_endpoint, self.refresh_token, self.client.client_id,
            self.client.client_secret, self.client.scope)
        response = requests.post(url)
        data = response.json()
        self.token_type = data['token_type']
        self.expires_in = data['expires_in']
        self.access_token = data['access_token']
        if 'refresh_token' in data:
            self.refresh_token = data['refresh_token']
        self.save()

    def expiration(self):
        "Returns the expiration date in the local timezone."
        modified = timezone.localtime(self.modified)
        delta = datetime.timedelta(seconds=self.expires_in)
        return modified + delta

    def is_expired(self):
        "Returns if the token is expired."
        return timezone.now() > self.expiration()
